### PR TITLE
Asynchronous patterns for Cloud Functions

### PR DESCRIPTION
**Pull request by ptone**
Focuses on Asynchronous Stateful Jobs

### CODE REVIEW

**Comment by ptone** [comment]
More readable review content:
https://github.com/ptone/community/blob/async-func/tutorials/cloud-functions-async/index.md

File: tutorials/cloud-functions-async/index.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
+---
+title: Asynchronous patterns for Cloud Functions
+description: Build a simple HTTP function API for long running job or tasks.
+author: ptone
+tags: Cloud Functions, Firestore, serverless
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: Firestore -> Cloud Firestore, here and any other mentions

File: tutorials/cloud-functions-async/index.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+### Asynchronous Stateful Jobs
+![](image/stateful-job.png)
+The caller does not want to wait for the work to be done, but would like to be able to inquire as to the completion of the request.
+
+This pattern is common enough that Google has defined a standard of a [long-running-operation API contract](https://github.com/googleapis/googleapis/tree/master/google/longrunning) used in multiple APIs.
+
+This is a basic pattern which may have some use-case specific details.
+
+## Serverless Stateful Jobs
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: This is a little confusing. I think this is an implementation of the 3rd use case identified above (Asynchronous Stateful Jobs) but I'm not certain. Can you add some clarification in the intro paragraph to this section regarding how the example in this tutorial maps to the use cases identified?

File: tutorials/cloud-functions-async/index.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+This tutorial uses several managed services to implement this pattern including [PubSub](https://cloud.google.com/pubsub/) and [Firestore](https://cloud.google.com/firestore/).
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: Clarify what "this pattern" refers to. Serverless stateful jobs, I think, but should be explicit.

Also, reference to PubSub should be "Cloud PubSub", here and throughout.

File: tutorials/cloud-functions-async/index.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+![](image/arch.png)
+
+You can implement job state storage or a work queue with alternatives like Cloud SQL and Cloud Tasks. This tutorial uses the support of [golang](https://golang.org/) in Cloud Functions, but any supported language can be used to fulfill the pattern.
+
+### Job resource definition
+
+A simple custom job definition is defined:
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: A simple custom job definition is defined as follows:

File: tutorials/cloud-functions-async/index.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	DoneTime   time.Time `json:"done-time" firestore:"DoneTime,omitempty"`
+	Done       bool
+	Result     string
+	State      JobState
+	Task       map[string]interface{} `json:"-"`
+	// can add a requester, source IP etc if needed
+}
+```
+
+When a job request is received, it is given an ID and the details of the work are included in a task field. This task is sent into the work queue as PubSub payload, with the Job ID as an [PubSub attribute](https://cloud.google.com/pubsub/docs/publisher#custom-attributes). When a worker picks up the task, it moves the job from `Created` -> `Running`. When the task is complete, it moves from `Running` to `Completed` (success) or `Failed` and writes any result back into the Firestore document.
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: PubSub payload -> a Cloud PubSub payload

with the Job ID as an [PubSub attribute] -> with the Job ID as a Cloud [PubSub attribute]

`Created` -> `Running` -> `Created` to `Running`

File: tutorials/cloud-functions-async/index.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+`gcloud functions deploy Jobs --runtime go111 --trigger-http`
+
+## Create Jobs and check their state
+
+Once the function is deployed, we can use both POST and GET methods to interact with long running jobs.
+
+### Get URL
+
+Get the URL of the deployed function into an environment variable:
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: Get the URL -> Copy the URL

File: tutorials/cloud-functions-async/index.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+date_published: 2019-02-20
+---
+
+Preston Holmes | Solution Architect | Google
+
+<!-- diagram sources: https://docs.google.com/presentation/d/1s01eqo3YUKiskJwSESW-T17IeUQf3DLCT_lvuAV7CwM/edit#slide=id.g4fb0d7b3af_0_0 -->
+
+## Introduction
+
+This tutorial demonstrates how to use [Google Cloud Functions](https://cloud.google.com/functions/) to extend the synchronous web-hook style request/response to longer running jobs with a focus on trackable and stateful long running operations.
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: longer running jobs -> longer running jobs,

File: tutorials/cloud-functions-async/index.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+Get the URL of the deployed function into an environment variable:
+
+```
+export URL=$(gcloud functions describe Jobs --format='value(httpsTrigger.url)')
+```
+
+### Create a Job
+
+Call the function based API to create a Job with a simple task. Our task has one field which is `worktime` which determines how many seconds the task takes to complete.
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: function based -> function-based

Our task has one field which is `worktime` which determines how many seconds the task takes to complete. - > Our task has one field, `worktime`, which determines how many seconds the task takes to complete.

File: tutorials/cloud-functions-async/index.md [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+```
+JOBID=$(curl -s --header "Content-Type: application/json"   --request POST   --data '{"worktime":40}'   $URL | jq -r '.id')
+```
+
+When the job is created, the function returns the ID of the job, which is put into an environment variable for reference later.
+
+### Check on Job status
+This command will poll the state of the job every 2 seconds with the [watch command](https://linux.die.net/man/1/watch).
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: This command will poll -> This command polls

File: tutorials/cloud-functions-async/index.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+```
+watch -t "curl -s $URL/$JOBID | python -m json.tool"
+```
+
+### Perform the work
+
+Start a worker in another tab in either your local terminal or in Cloud Shell and run a basic worker that will complete after waiting the `worktime` value in the original
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: Start a worker in another tab in either your local terminal or in Cloud Shell and run a basic worker that will complete after waiting the `worktime` value in the original  -> Start a worker in another tab in either your local terminal or in Cloud Shell, and then run a basic worker that will complete after waiting the `worktime` value in the original.

File: tutorials/cloud-functions-async/index.md [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+Start a worker in another tab in either your local terminal or in Cloud Shell and run a basic worker that will complete after waiting the `worktime` value in the original 
+
+```
+cd worker
+go run main.go
+```
+
+## Cleanup and Next steps
+
+You can delete the deployed function and PubSub resources:
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: I'd suggest either "Optionally, delete the deployed function and PubSub resources:" or "Delete the deployed function and PubSub resources:", depending on whether you want to leave them the option of keeping resources to use as a starting point for production code.

**Comment by ptone** [comment]
Thanks @ccarpentiere for your review comments, I also got some golang code comments in a side channel and will try to address all of this in the next couple days, just wanted to acknowledge your help.

File: tutorials/cloud-functions-async/index.md (hunk 1 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: My understanding that we can drop "Cloud" after initial occurrence?

I did add it to all occurrences in headings and initial use

File: tutorials/cloud-functions-async/index.md (hunk 2 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: done

File: tutorials/cloud-functions-async/index.md (hunk 3 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: done

File: tutorials/cloud-functions-async/index.md (hunk 4 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: done

File: tutorials/cloud-functions-async/index.md (hunk 1 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: Hmm - looking at https://cloud.google.com/guides/style/branding-guidelines (internal link) seems like Cloud is now included in all "short names" as well  - so addressing this.

File: tutorials/cloud-functions-async/index.md (hunk 5 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: done

File: tutorials/cloud-functions-async/index.md (hunk 6 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: done

File: tutorials/cloud-functions-async/index.md (hunk 7 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: done

File: tutorials/cloud-functions-async/index.md (hunk 8 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: done

File: tutorials/cloud-functions-async/index.md (hunk 9 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: done

File: tutorials/cloud-functions-async/index.md (hunk 10 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: done

**Comment by ptone** [comment]
I'm going to run through again in a clean project after addressing some technical feedback, writing content ready for re-review

**Comment by ptone** [comment]
I've run though again in a clean project and fixed remaining tech glitches I believe

cc @halvards

File: tutorials/cloud-functions-async/index.md [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+There are a couple ways this can be solved on Google Cloud Platform in a serverless way. You can use Cloud Pub/Sub patterns for [long running tasks](https://cloud.google.com/solutions/using-cloud-pub-sub-long-running-tasks) or you can use a dedicated service with [Cloud Tasks](https://cloud.google.com/tasks/).
+
+### Asynchronous Stateful Jobs
+![](image/stateful-job.png)
+The caller does not want to wait for the work to be done, but would like to be able to inquire as to the completion of the request.
+
+This pattern is common enough that Google has defined a standard of a [long-running-operation API contract](https://github.com/googleapis/googleapis/tree/master/google/longrunning) used in multiple APIs.
+
+This is a high level pattern. What about a job state is tracked and how work is performed will vary by use-case. This tutorial dives deeper into this pattern
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: dives deeper into this pattern.

File: tutorials/cloud-functions-async/index.md [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+
+## Serverless Stateful Jobs
+
+This tutorial uses several managed services to implement the asyn c stateful jobs pattern including [Cloud Pub/Sub](https://cloud.google.com/pubsub/) and [Cloud Firestore](https://cloud.google.com/firestore/).
```
Comment (ccarpentiere) [inline tutorials/cloud-functions-async/index.md]: asyn c -> async

File: tutorials/cloud-functions-async/index.md (hunk 12 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: fixed

File: tutorials/cloud-functions-async/index.md (hunk 13 above)
Comment (ptone) [reply tutorials/cloud-functions-async/index.md]: fixed

**Comment by ptone** [comment]
@ToddKopriva this is ready for merge and publish - thx

**Comment by ToddKopriva** [comment]
I'll do my pass on it by EOD tomorrow.

**Comment by ToddKopriva** [comment]
https://cloud.google.com/community/tutorials/cloud-functions-async

**Pull request merged**